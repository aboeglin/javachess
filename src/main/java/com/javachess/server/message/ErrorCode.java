package com.javachess.server.message;

public enum ErrorCode {
  MOVE_NOT_ALLOWED,
  NOT_YOUR_TURN,
  PLAYER_NOT_IN_GAME,
  GAME_ALREADY_FULL,
  GAME_NOT_FOUND,
}
