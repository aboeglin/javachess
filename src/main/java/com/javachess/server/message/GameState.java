package com.javachess.server.message;

import com.javachess.logic.Game;
import com.javachess.logic.Piece;
import com.javachess.logic.Player;

import java.util.List;

public class GameState {

  private int id;
  private Player player1;
  private Player player2;
  private Player activePlayer;
  private List<Piece> pieces;

  public GameState(Game g) {
    this.id = g.getId();
    this.player1 = g.getPlayer1();
    this.player2 = g.getPlayer2();
    this.activePlayer = Game.getActivePlayer(g);
    this.pieces = Game.getPieces(g);
  }

}
