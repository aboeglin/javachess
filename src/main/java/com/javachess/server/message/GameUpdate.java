package com.javachess.server.message;

import com.javachess.logic.Game;
import com.javachess.logic.Piece;
import com.javachess.logic.Player;
import com.javachess.logic.Position;

import java.util.List;

public class GameUpdate {

  private String status;

  private GameState game;

  private boolean error;

  private ErrorCode errorCode;

  private String errorMessage;

  private Position[] possibleMoves;

  public GameUpdate(String status, Game game) {
    this.status = status;
    this.game = new GameState(game);
    this.possibleMoves = new Position[]{};
  }

  public GameUpdate(String status, Game game, Position[] possibleMoves) {
    this.status = status;
    this.game = new GameState(game);
    this.possibleMoves = possibleMoves;
  }

  public GameUpdate(String status, Game game, ErrorCode errorCode, String errorMessage) {
    this.status = status;
    this.game = new GameState(game);
    this.errorMessage = errorMessage;
    this.errorCode = errorCode;
    this.error = true;
    this.possibleMoves = new Position[]{};
  }
}

class GameState {

  private int id;
  private Player player1;
  private Player player2;
  private Player activePlayer;
  private List<Piece> pieces;

  public GameState(Game g) {
    this.id = g.getId();
    this.player1 = g.getPlayer1();
    this.player2 = g.getPlayer2();
    this.activePlayer = Game.getActivePlayer(g);
    this.pieces = Game.getPieces(g);
  }

}
