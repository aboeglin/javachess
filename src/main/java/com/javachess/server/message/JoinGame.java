package com.javachess.server.message;

public class JoinGame {
  private String playerId;

  public JoinGame(String playerId) {
    this.playerId = playerId;
  }

  public String getPlayerId() {
    return playerId;
  }
}
