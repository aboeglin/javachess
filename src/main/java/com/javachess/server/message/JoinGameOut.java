package com.javachess.server.message;

public class JoinGameOut {
  private String message;

  public JoinGameOut(String message) {
    this.message = message;
  }

  public String getMessage() {
    return message;
  }

  public void setMessage(String message) {
    this.message = message;
  }
}
