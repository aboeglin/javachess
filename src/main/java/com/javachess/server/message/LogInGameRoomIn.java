package com.javachess.server.message;

public class LogInGameRoomIn {
  private String email;

  public LogInGameRoomIn(String email) {
    this.email = email;
  }

  public String getEmail() {
    return email;
  }

  public void setEmail(String email) {
    this.email = email;
  }
}
