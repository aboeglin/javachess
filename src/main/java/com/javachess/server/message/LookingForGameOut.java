package com.javachess.server.message;

public class LookingForGameOut {

  private String message;

  public LookingForGameOut(String message) {
    this.message = message;
  }

  public String getMessage() {
    return message;
  }

  public void setMessage(String message) {
    this.message = message;
  }
}
