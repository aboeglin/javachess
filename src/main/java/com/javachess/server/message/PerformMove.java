package com.javachess.server.message;

public class PerformMove {

  private String playerId;
  private String fromX;
  private String fromY;
  private String toX;
  private String toY;

  public PerformMove(
    String playerId,
    String fromX, String fromY,
    String toX, String toY
  ) {
    this.playerId = playerId;
    this.fromX = fromX;
    this.fromY = fromY;
    this.toX = toX;
    this.toY = toY;
  }

  public String getPlayerId() {
    return playerId;
  }

  public void setPlayerId(String playerId) {
    this.playerId = playerId;
  }

  public String getFromX() {
    return fromX;
  }

  public void setFromX(String fromX) {
    this.fromX = fromX;
  }

  public String getFromY() {
    return fromY;
  }

  public void setFromY(String fromY) {
    this.fromY = fromY;
  }

  public String getToX() {
    return toX;
  }

  public void setToX(String toX) {
    this.toX = toX;
  }

  public String getToY() {
    return toY;
  }

  public void setToY(String toY) {
    this.toY = toY;
  }
}
