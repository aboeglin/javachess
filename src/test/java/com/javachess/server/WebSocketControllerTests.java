package com.javachess.server;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.PrettyPrinter;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.core.util.Separators;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.google.gson.Gson;
import com.javachess.server.message.JoinGameIn;
import com.javachess.server.message.JoinGameOut;
import com.javachess.server.message.LookingForGameOut;
import io.github.jsonSnapshot.SnapshotMatcher;
import org.junit.jupiter.api.*;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.messaging.converter.StringMessageConverter;
import org.springframework.messaging.simp.stomp.*;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.web.socket.client.standard.StandardWebSocketClient;
import org.springframework.web.socket.messaging.WebSocketStompClient;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.UncheckedIOException;
import java.lang.reflect.Field;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class WebSocketControllerTests {

  @Value("${local.server.port}")
  private int port;

  private WebSocketStompClient client;
  private StompSession session;

  @BeforeAll
  public static void beforeAll() {
    SnapshotMatcher.start(Snapshot::asJsonString);
  }

  @AfterAll
  public static void afterAll() {
    SnapshotMatcher.validateSnapshots();
  }

  @BeforeEach
  public void setUp() throws Exception {
    String url = String.format("ws://localhost:%s/ws", this.port);

    this.client = new WebSocketStompClient(new StandardWebSocketClient());
    this.client.setMessageConverter(new StringMessageConverter());
    this.session = this.client.connect(url, new MyStompSessionHandler()).get();
  }

  @AfterEach
  public void tearDown() throws Exception {
    this.session.disconnect();
    this.client.stop();
  }

  @Test
  public void connectsToSocket() throws Exception {
    assertEquals(this.session.isConnected(), true);
  }

  @Test
  @DisplayName("It should respond with a game id")
  public void lookForGame() throws Exception {
    CompletableFuture<String> resultKeeper = new CompletableFuture<>();

    this.session.subscribe("/user/queue/lfg/ack", new TestStompFrameHandler(payload -> {
      resultKeeper.complete(payload);
    }));
    Thread.currentThread().sleep(1000);

    this.session.send("/app/lfg", "{email: test}");

    assertEquals(resultKeeper.get(2, TimeUnit.SECONDS), "{\"message\":\"Hi test, looking for opponent ...\",\"gameId\":1}");
  }

  @Test
  @DisplayName("It respond on the endpoint of the game room")
  public void joinRoom() throws Exception {
    CompletableFuture<String> lfgKeeper = new CompletableFuture<>();
    CompletableFuture<String> resultKeeper = new CompletableFuture<>();

    this.session.subscribe("/user/queue/lfg/ack", new TestStompFrameHandler(payload ->
      lfgKeeper.complete(payload)
    ));
    Thread.currentThread().sleep(1000);

    this.session.send("/app/lfg", "{email: test}");

    Gson gson = new Gson();
    LookingForGameOut response = gson.fromJson(lfgKeeper.get(2, TimeUnit.SECONDS), LookingForGameOut.class);

    this.session.subscribe(String.format("/queue/game/%s/ready", response.getGameId()), new TestStompFrameHandler(payload ->
      resultKeeper.complete(payload)
    ));

    this.session.send(String.format("/app/game/%s/join", response.getGameId()), "{email: test}");
    SnapshotMatcher.expect(resultKeeper.get(2, TimeUnit.SECONDS)).toMatchSnapshot();
  }
}

class TestStompFrameHandler implements StompFrameHandler {

  private final Consumer<String> frameHandler;

  public TestStompFrameHandler(Consumer<String> c) {
    this.frameHandler = c;
  }

  @Override
  public Type getPayloadType(StompHeaders stompHeaders) {
    return String.class;
  }

  @Override
  public void handleFrame(StompHeaders stompHeaders, Object payload) {
    this.frameHandler.accept(payload.toString());
  }
}

class MyStompSessionHandler extends StompSessionHandlerAdapter {
  @Override
  public void afterConnected(StompSession session, StompHeaders connectedHeaders) {
    super.afterConnected(session, connectedHeaders);
  }

  @Override
  public void handleException(StompSession session, StompCommand command, StompHeaders headers, byte[] payload, Throwable exception) {
    super.handleException(session, command, headers, payload, exception);
  }
}

final class Snapshot {
  private static final ObjectMapper objectMapper = buildObjectMapper();
  private static final PrettyPrinter pp = buildDefaultPrettyPrinter();

  /**
   * Workaround for an incompatibility between latest Jackson and json-snapshot libs.
   * <p>
   * Intended to replace {@code io.github.jsonSnapshot.SnapshotMatcher#defaultJsonFunction}
   *
   * @see <a href="https://github.com/json-snapshot/json-snapshot.github.io/issues/27">Issue in json-snapshot project</a>
   */
  public static String asJsonString(Object object) {
    try {
      return objectMapper.writer(pp).writeValueAsString(object);
    } catch (JsonProcessingException e) {
      throw new UncheckedIOException(e);
    }
  }

  /**
   * Unmodified copy of {@code io.github.jsonSnapshot.SnapshotMatcher#buildObjectMapper}
   */
  private static ObjectMapper buildObjectMapper() {
    ObjectMapper objectMapper = new ObjectMapper();
    objectMapper.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
    objectMapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);

    objectMapper.setVisibility(
      objectMapper
        .getSerializationConfig()
        .getDefaultVisibilityChecker()
        .withFieldVisibility(JsonAutoDetect.Visibility.ANY)
        .withGetterVisibility(JsonAutoDetect.Visibility.NONE)
        .withSetterVisibility(JsonAutoDetect.Visibility.NONE)
        .withCreatorVisibility(JsonAutoDetect.Visibility.NONE));
    return objectMapper;
  }

  /**
   * Modified copy of {@code io.github.jsonSnapshot.SnapshotMatcher#buildDefaultPrettyPrinter}
   */
  private static PrettyPrinter buildDefaultPrettyPrinter() {
    DefaultPrettyPrinter pp =
      new DefaultPrettyPrinter("") {
        @Override
        public DefaultPrettyPrinter createInstance() {
          return this;
        }

        @Override
        public DefaultPrettyPrinter withSeparators(Separators separators) {
          this._separators = separators;
          this._objectFieldValueSeparatorWithSpaces =
            separators.getObjectFieldValueSeparator() + " ";
          return this;
        }
      };
    DefaultPrettyPrinter.Indenter lfOnlyIndenter = new DefaultIndenter("  ", "\n");
    pp.indentArraysWith(lfOnlyIndenter);
    pp.indentObjectsWith(lfOnlyIndenter);
    return pp;
  }
}