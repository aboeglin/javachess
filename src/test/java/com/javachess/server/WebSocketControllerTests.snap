com.javachess.server.WebSocketControllerTests.joinRoom=[
  "{\"message\":\"In the room !\nThe game will start shortly ...\"}"
]